rtl/eclrtl/eclhelper_dyn.cpp

```diff
@@ -40,63 +40,167 @@
+    else if (*a==*b)
+        return 0;
+    else
+        return 1;
+}
+
+class FilterSet
+{
+public:
+    FilterSet(const RtlRecord &_inrec) : inrec(_inrec)
+    {
+
+    }
+    void addFilter(const char *filter)
+    {
+        // Format of a filter is:
+        // field[..n]: valuestring
+        // value string format specifies ranges using a comma-separated list of ranges.
+        // Each range is specified as paren lower, upper paren, where the paren is either ( or [ depending
+        // on whether the specified bound is inclusive or exclusive.
+        // If only one bound is specified then it is used for both upper and lower bound (only meaningful with [] )
+        //
+        // ( A means values > A - exclusive
+        // [ means values >= A - inclusive
+        // A ) means values < A - exclusive
+        // A ] means values <= A - inclusive
+        // For example:
+        // [A] matches just A
+        // (,A),(A,) matches all but A
+        // (A] of [A) are both empty ranges
+        // [A,B) means A*
+        // Values use the ECL syntax for constants. String constants are always utf8. Binary use d'xx' format (hexpairs)
+        // Note that binary serialization format is different
+
+        assertex(filter);
+        const char *epos = strpbrk(filter,"=~");
+        if (!epos)
+            throw MakeStringException(0, "Invalid filter string: expected = or ~ after fieldname");
+        StringBuffer fieldName(epos-filter, filter);
+        unsigned fieldNum = inrec.getFieldNum(fieldName);
+        if (fieldNum == (unsigned) -1)
+            throw MakeStringException(0, "Invalid filter string: field '%s' not recognized", fieldName.str());
+        unsigned numOffsets = inrec.getNumVarFields() + 1;
+        size_t * variableOffsets = (size_t *)alloca(numOffsets * sizeof(size_t));
+        RtlRow offsetCalculator(inrec, nullptr, numOffsets, variableOffsets);
+        unsigned fieldOffset = offsetCalculator.getOffset(fieldNum);
+        unsigned fieldSize = offsetCalculator.getSize(fieldNum);
+        const RtlTypeInfo *fieldType = inrec.queryType(fieldNum);
+        filter = epos+1;
+        if (*filter=='~')
+        {
+            UNIMPLEMENTED;  // use a regex?
+        }
+        else
+        {
+            MemoryBuffer lobuffer;
+            MemoryBuffer hibuffer;
+            Owned<IStringSet> filterSet = createStringSet(fieldSize);
+            deserializeSet(*filterSet, inrec.getMinRecordSize(), fieldType, filter);
+            aindex_t found = filterOffsets.bSearch(fieldOffset, compareOffsets);
+            if (found==NotFound)
+            {
+                bool isNew;
+                unsigned insertPos = filterOffsets.bAdd(fieldOffset, compareOffsets, isNew);
+                dbgassertex(isNew);
+                filters.add(*filterSet.getClear(), insertPos);
+            }
+            else
+            {
+                filters.item(found).unionSet(filterSet);  // Debatable - would intersect be more appropriate?
```

Comment:
The old IStringSets unionSet returns the result - rather than modifying the set => this will leak and be a nop.  (New sets modify in place.)

Reply:
Will fix